### PR TITLE
Enhance telemetry information.

### PR DESCRIPTION
**Pull request by cristinamanum**
## Bug

https://github.com/NuGet/Home/issues/10067

## Fix

Details: Adds new telemetry information.
Adds 2 new overloads for TelemetryActivity: StartIntervalMeasure and EndIntervalMeasure

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  covered through all other scenarios
Validation:  Manual validation

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 1]
```diff
@@ -209,6 +218,19 @@ public static async Task<IReadOnlyList<RestoreSummaryRequest>> GetRequests(Resto
             return requests;
         }
 
+        /// <summary>
+        /// If the request does not have a ParentId add one.
+        /// This will enable event telemetry correlation.
+        /// </summary>
+        /// <param name="request">The request</param>
+        private static void AddParentIdToRequest(RestoreSummaryRequest request)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs]: When is this generated/not generated today? 

Can we just generate the parent id at the beginning when creating this instead of populating it later.
Populating missing data can be hard to follow sometimes imho.

File: src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1,2 +1,4 @@
 NuGet.Common.NuGetLogCode.NU1010 = 1010 -> NuGet.Common.NuGetLogCode
 NuGet.Common.NuGetLogCode.NU1011 = 1011 -> NuGet.Common.NuGetLogCode
+NuGet.Common.TelemetryActivity.EndIntervalMeasure(string propertyName, System.Diagnostics.Stopwatch stopWatch) -> void
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt:3]: What's the motivation here?
Because non overlapping ranges are not supported? 

Might be a good idea to move this to Common then?

https://github.com/NuGet/NuGet.Client/blob/ef00db94960b467d8318c8b3cd1760861c567eff/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/IntervalTracker.cs#L17

I'd wait on @zivkan's feedback for that though :)

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs [hunk 3]
```diff
@@ -14,7 +15,7 @@ public class TelemetryActivity : IDisposable
         private readonly DateTime _startTime;
         private readonly Stopwatch _stopwatch;
         private readonly Stopwatch _intervalWatch = new Stopwatch();
-        private readonly List<Tuple<string, TimeSpan>> _intervalList;
+        private readonly ConcurrentBag<Tuple<string, TimeSpan>> _intervalList;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:18]: Why do we need a concurrent bag here?
Is there a concurrency problem? Is this new?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs [hunk 4]
```diff
@@ -57,6 +57,8 @@ public class RestoreCommand
         private const string MsbuildAssetsVerificationDuration = "MsbuildAssetsVerificationDuration";
         private const string MsbuildAssetsVerificationResult = "MsbuildAssetsVerificationResult";
         private const string ReplayLogsDuration = "ReplayLogsDuration";
+        private const string EvaluateCacheFileGetHashDuration = "EvaluateCacheFileGetHashDuration";
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Can we remove the `EvaluateCacheFile` part out of the name?
The interval names don't have to match exact methods, but rather the concepts. 
So how about `GetNoOpHashDuration` and `CacheFileReadDuration`?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs (hunk 1 above)
Comment (cristinamanum) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs]: Currently in two random days of data there are 0.05% of devenv projectrestoreinformation events without a parent id. I think that this addition may not be needed. I reverted the changes.

File: src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (cristinamanum) [reply src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt:3]: How the NuGetCore events can use the IntervalTracker?

Comment (cristinamanum) [reply src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt:3]: > What's the motivation here?
> Because non overlapping ranges are not supported?

Yes

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs (hunk 3 above)
Comment (cristinamanum) [reply src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:18]: The current pattern of using the interval list is sequential
StartIntervalMeasure()
// do some work
EndIntervalMeasure()

With new proposed API 
``` csharp
 public void EndIntervalMeasure(string propertyName, Stopwatch stopWatch)
```

the list _intervalList could be updated concurrently.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs (hunk 4 above)
Comment (cristinamanum) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Yes, updated.

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:18]: Can we use a simple object lock? 

It should be more than sufficient because a ConcurrentBag is really optimized for read/write scenarios.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:18]: I think we can move https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/IntervalTracker.cs all the way to NuGet.Common because it's a scenario that we'll need as we try to reduce the number of events NuGet is sending. 

We might need to remove the Etw part of it though :) Hence why I called for @zivkan's opinion :)

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 5]
```diff
@@ -228,9 +235,21 @@ private static int GetMaxTaskCount(RestoreArgs restoreContext)
 
         private static async Task<RestoreSummary> ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
         {
-            var result = await ExecuteAsync(summaryRequest, token);
+            using (var telemetry = TelemetryActivity.Create(parentId: summaryRequest.Request.ParentId, eventName: ProjectExecuteAndCommitAsyncInformation))
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:238]: I'm not excited about adding yet another event in this codepath. 
https://github.com/NuGet/Client.Engineering/issues/319 calls for consolidating the restore telemetry into one event. I don't necessarily want to block this PR ofc, because it's easy to refactor around it, there's no customer contract, but something to keep in mind.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:238]: The solution might be starting the event *here* and then just the restore command adds measurements.

**Review by nkolev92 — COMMENTED** [review]
In general the PR is good, just want to get a pair of eyes on the new API. 

A couple of things: 

* PR body has a link to the issue, but doesn't have a fixes statement. This means that the issue won't get closed. Lots of people mess with the PR template, so I think it's definitely in need of improvement. 

* PR title doesn't tell us much. It doesn't tell us which telemetry, for what component, what command, what particular part is enhanced. The title should focus on the how.

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs (hunk 3 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:18]: I think `IntervalTracker` only makes sense in code that is instrumented to emit telemetry.  If we go towards my goal of raising ETW events in the shared code, and listen to those ETW events in the VS code and transform it into a telemetry event, then there might not be a use for `IntervalTracker` in `NuGet.Common`. It would just be a wrapper for `Stopwatch`, without any real benefit, right?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:18]: It allows you to have an easy "Start/Stop" pattern, where it's a disposable instead of having to create your own stopwatch to pass to the TelemetryActivity. 

I think it makes sense for tracking internals :) Whether you send telemetry is up to the user.

File: src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt:3]: I don't think that `EndIntervalMeasure` is a good API name, because it has nothing to do with `StartIntervalMeasure`. `AddIntervalMeasure` would be more appropriate.

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs [hunk 6]
```diff
@@ -80,6 +81,22 @@ public void EndIntervalMeasure(string propertyName)
...
+        {
+            stopWatch.Restart();
+        }
+
+        /// <summary> End interval measure. </summary>
+        /// <param name="propertyName"> Property name to represents the interval. </param>
+        /// <param name="stopWatch"> Stopwatch used to measure the time interval. </param>
+        public void EndIntervalMeasure(string propertyName, Stopwatch stopWatch)
+        {
+            stopWatch.Stop();
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:96]: The `TelemetryActivity` class didn't start the stopwatch passed in, so I don't think it should stop it. It's not the "owner" of the stopwatch.

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs [hunk 7]
```diff
@@ -80,6 +81,22 @@ public void EndIntervalMeasure(string propertyName)
...
+        /// <param name="stopWatch"> Stopwatch used to measure the time interval. </param>
+        public void StartIntervalMeasure(Stopwatch stopWatch)
+        {
+            stopWatch.Restart();
+        }
+
+        /// <summary> End interval measure. </summary>
+        /// <param name="propertyName"> Property name to represents the interval. </param>
+        /// <param name="stopWatch"> Stopwatch used to measure the time interval. </param>
+        public void EndIntervalMeasure(string propertyName, Stopwatch stopWatch)
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:94]:
```suggestion
        public void EndIntervalMeasure(string propertyName, Stopwatch stopwatch)
```

File: src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs [hunk 8]
```diff
@@ -80,6 +81,22 @@ public void EndIntervalMeasure(string propertyName)
             _intervalList.Add(new Tuple<string, TimeSpan>(propertyName, _intervalWatch.Elapsed));
         }
 
+        /// <summary> Start interval measure. </summary>
+        /// <param name="stopWatch"> Stopwatch used to measure the time interval. </param>
+        public void StartIntervalMeasure(Stopwatch stopWatch)
+        {
+            stopWatch.Restart();
+        }
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Common/Telemetry/TelemetryActivity.cs:89]: I don't think this is a good API. This class `TelemetryActivity` doesn't "own" the stopwatch. This API doesn't provide the caller any guarantees, it's a trivial wrapper of `stopWatch.Resart`.

**Comment by nkolev92** [comment]
Closing the PR for now, as it's not an immediate priority.
